File: hop_reporter.py
#!/usr/bin/env python3
"""Companion hop scanner — continuously reports what this machine hears to the hop graph."""

from __future__ import annotations

import argparse
import asyncio
import json
import sys
import time
import urllib.error
import urllib.request

from bleak import BleakScanner

from ble_device_naming import format_mac


async def scan_once(duration: float) -> list[dict]:
    seen: dict[str, dict] = {}

    def callback(device, adv):
        addr = format_mac(device.address)
        name = adv.local_name or device.name
        seen[addr] = {
            "address": addr,
            "name": name,
            "rssi": adv.rssi,
            "seenAt": int(time.time() * 1000),
        }

    scanner = BleakScanner(detection_callback=callback, scanning_mode="active")
    await scanner.start()
    try:
        await asyncio.sleep(duration)
    finally:
        await scanner.stop()

    return list(seen.values())


def post_report(server: str, node_id: str, label: str, self_address: str | None, observations: list[dict], listening_post: bool = False) -> dict:
    payload = {
        "nodeId": node_id,
        "nodeLabel": label,
        "selfAddress": self_address,
        "observations": observations,
        "listeningPost": listening_post,
    }
    body = json.dumps(payload).encode("utf-8")
    req = urllib.request.Request(
        f"{server.rstrip('/')}/api/hop/report",
        data=body,
        headers={"Content-Type": "application/json"},
        method="POST",
    )
    with urllib.request.urlopen(req, timeout=15) as resp:
        return json.loads(resp.read().decode("utf-8"))


async def run_loop(
    server: str,
    node_id: str,
    label: str,
    self_address: str | None,
    duration: float,
    interval: float,
    listening_post: bool,
) -> None:
    tag = "LISTENING POST" if listening_post else "hop node"
    print(f"Continuous {tag} '{label}' → {server} (scan {duration}s, repeat every {interval}s)")
    while True:
        observations = await scan_once(duration)
        depth = 0
        try:
            result = post_report(server, node_id, label, self_address, observations, listening_post)
            depth = result.get("hopGraph", {}).get("maxHopDepth", 0)
            print(
                f"[{time.strftime('%H:%M:%S')}] reported {len(observations)} contact(s) · "
                f"graph depth {depth}"
            )
        except urllib.error.URLError as exc:
            print(f"[{time.strftime('%H:%M:%S')}] server unreachable: {exc}", file=sys.stderr)
        await asyncio.sleep(interval)


async def main() -> int:
    parser = argparse.ArgumentParser(description="BLE hop companion reporter")
    parser.add_argument("--server", default="http://127.0.0.1:8765", help="Hop graph server URL")
    parser.add_argument("--node-id", required=True, help="Unique scanner id (e.g. pixel-hop-1)")
    parser.add_argument("--label", default=None, help="Human label for this scanner")
    parser.add_argument("--self-address", default=None, help="This device's BLE MAC (links domino chain)")
    parser.add_argument("--duration", type=float, default=12.0, help="Seconds to scan each cycle")
    parser.add_argument("--interval", type=float, default=15.0, help="Seconds between hop reports (loop mode)")
    parser.add_argument("--loop", action="store_true", help="Never stop — continuous domino hop reporting")
    parser.add_argument("--listening-post", action="store_true", help="Register as fixed LISTENING POST dead-drop node")
    args = parser.parse_args()

    label = args.label or args.node_id

    if args.loop:
        await run_loop(args.server, args.node_id, label, args.self_address, args.duration, args.interval, args.listening_post)
        return 0

    print(f"Scanning {args.duration}s as hop node '{label}'...")
    observations = await scan_once(args.duration)
    print(f"Seen {len(observations)} device(s), posting to {args.server}...")

    try:
        result = post_report(args.server, args.node_id, label, args.self_address, observations, args.listening_post)
        print(json.dumps(result, indent=2))
        return 0
    except urllib.error.URLError as exc:
        print(f"Failed to reach server: {exc}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    raise SystemExit(asyncio.run(main()))

File: src/cli/hop-reporter.ts
#!/usr/bin/env node
/** Companion hop scanner — reports observations to root mapper. */

import { parseArgs } from "node:util";
import { formatMac } from "../ble/device-naming.js";
import { scanOnce } from "../server/scanner.js";

async function postReport(
  server: string,
  nodeId: string,
  label: string,
  selfAddress: string | null,
  observations: Record<string, unknown>[],
  listeningPost = false,
  latitude?: number,
  longitude?: number,
  accuracyMeters?: number,
): Promise<Record<string, unknown>> {
  const payload: Record<string, unknown> = {
    nodeId,
    nodeLabel: label,
    selfAddress,
    observations,
    listeningPost,
  };
  if (latitude != null && longitude != null) {
    payload.latitude = latitude;
    payload.longitude = longitude;
    if (accuracyMeters != null) payload.accuracyMeters = accuracyMeters;
  }
  const res = await fetch(`${server.replace(/\/$/, "")}/api/hop/report`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  const data = (await res.json()) as Record<string, unknown> & { error?: string };
  if (!res.ok) throw new Error(data.error ?? res.statusText);
  return data;
}

async function runLoop(opts: {
  server: string;
  nodeId: string;
  label: string;
  selfAddress: string | null;
  duration: number;
  interval: number;
  listeningPost: boolean;
  latitude?: number;
  longitude?: number;
  accuracyMeters?: number;
}): Promise<void> {
  const tag = opts.listeningPost ? "LISTENING POST" : "hop node";
  console.log(
    `Continuous ${tag} '${opts.label}' → ${opts.server} (scan ${opts.duration}s, repeat every ${opts.interval}s)`,
  );
  while (true) {
    const observations = await scanOnce(opts.duration);
    try {
      const result = await postReport(
        opts.server,
        opts.nodeId,
        opts.label,
        opts.selfAddress,
        observations,
        opts.listeningPost,
        opts.latitude,
        opts.longitude,
        opts.accuracyMeters,
      );
      const depth = ((result.hopGraph as Record<string, unknown>)?.maxHopDepth as number) ?? 0;
      console.log(
        `[${new Date().toLocaleTimeString()}] reported ${observations.length} contact(s) · graph depth ${depth}`,
      );
    } catch (e) {
      console.error(`[${new Date().toLocaleTimeString()}] server unreachable:`, e);
    }
    await new Promise((r) => setTimeout(r, opts.interval * 1000));
  }
}

async function main(): Promise<number> {
  const { values } = parseArgs({
    options: {
      server: { type: "string", default: "http://127.0.0.1:8765" },
      "node-id": { type: "string" },
      label: { type: "string" },
      "self-address": { type: "string" },
      duration: { type: "string", default: "12" },
      interval: { type: "string", default: "15" },
      loop: { type: "boolean", default: false },
      "listening-post": { type: "boolean", default: false },
      latitude: { type: "string" },
      longitude: { type: "string" },
      "accuracy-meters": { type: "string" },
    },
  });

  const nodeId = values["node-id"];
  if (!nodeId) {
    console.error("--node-id is required");
    return 1;
  }
  const label = values.label ?? nodeId;
  const selfAddress = values["self-address"] ? formatMac(values["self-address"]) : null;
  const duration = Number(values.duration);
  const interval = Number(values.interval);
  const lat = values.latitude != null ? Number(values.latitude) : undefined;
  const lon = values.longitude != null ? Number(values.longitude) : undefined;
  const acc = values["accuracy-meters"] != null ? Number(values["accuracy-meters"]) : undefined;

  if (values.loop) {
    await runLoop({
      server: values.server!,
      nodeId,
      label,
      selfAddress,
      duration,
      interval,
      listeningPost: values["listening-post"] ?? false,
      latitude: lat,
      longitude: lon,
      accuracyMeters: acc,
    });
    return 0;
  }

  console.log(`Scanning ${duration}s as hop node '${label}'...`);
  const observations = await scanOnce(duration);
  console.log(`Seen ${observations.length} device(s), posting to ${values.server}...`);
  try {
    const result = await postReport(
      values.server!,
      nodeId,
      label,
      selfAddress,
      observations,
      values["listening-post"] ?? false,
      lat,
      lon,
      acc,
    );
    console.log(JSON.stringify(result, null, 2));
    return 0;
  } catch (e) {
    console.error("Failed to reach server:", e);
    return 1;
  }
}

main().then((code) => process.exit(code));
